### PR TITLE
Fix harming potion dupe

### LINKED ISSUE #3871: Potion of Harming Dupe
### What behaviour is expected:
Upon drinking a potion of harming that would cause lethal damage, the player drops an empty bottle.
(In regular Minecraft this is glitched as well, but differently. The bottle stays in the inventory on death and disappears completely on respawn. This isn't much better but at least it doesn't create a dupe.)

### What behaviour is observed:
If the players dies from drinking a potion of harming, they drop a duped potion of harming and keep the empty bottle in their inventory on death. When they respawn, they still have the bottle in their inventory.

### Steps/models to reproduce:
Bring your health down low enough and drink a potion of harming that would cause lethal damage.

### Plugin list:
EssentialsX
Multiverse/Multiverse-Inventories
Towny
Vault
WorldEdit
WorldGuard

### Paper build number:
git-Paper-79 (MC: 1.16.1) (API version 1.16.1-R0.1-SNAPSHOT)

### Anything else:
No additional information.

### PR DESCRIPTION
**Pull request by PepperCode1**
Fixes #3871

EntityLiving#applyInstantEffect() immediately kills the player and drops their inventory.
Before this patch, instant effects would be applied before the potion ItemStack is removed and replaced with a glass bottle. This caused the potion ItemStack to be dropped before it was supposed to be removed from the inventory. It also caused the glass bottle to be put into a dead player's inventory.
This patch makes it so that instant effects are applied after the potion ItemStack is removed, and the glass bottle is only put into the player's inventory if the player is not dead. Otherwise, the glass bottle is dropped on the ground.

### CODE REVIEW

File: Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch
Comment (JRoy) [inline Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch]:
```suggestion
```

File: Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch
Comment (JRoy) [inline Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch]:
```suggestion
+        List<MobEffect> instantLater = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(); // Paper - Fix harming potion dupe
```

**Review by JRoy — CHANGES_REQUESTED** [review]
don't change imports and use fastutil

**Comment by electronicboy** [comment]
Is this a vanilla issue or? is there is mojira ticket?

**Comment by PepperCode1** [comment]
This is a semi-Vanilla issue, as the linked Github issue states. In Vanilla, the potion is used up but the resulting glass bottle is deleted, while in Paper the potion turns into a glass bottle but drops an extra potion. I also can't find a bug report for this on Mojang, Mojira, or Spigot.

File: Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch
Comment (MiniDigger) [inline Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch]: this if seems unnecessary, last I checked paper code wasn't running on the client :P

File: Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch
Comment (MiniDigger) [inline Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch]: whats the purpose of the entityhuman null check here?

**Review by MiniDigger — DISMISSED** [review]
looks good, remove the import and the if (or well, we can handle that too once core team has approved) and its good to go

**Comment by PepperCode1** [comment]
I'll rebase, make the necessary changes, and do some testing.

File: Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch
Comment (PepperCode1) [reply Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch]: It has been removed in the latest version. I'm not sure why it was there either.

File: Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch
Comment (PepperCode1) [reply Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch]: The client side check has been removed in the latest version.

File: Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch
Comment (PepperCode1) [reply Spigot-Server-Patches/0549-Fix-harming-potion-dupe.patch]: The import has been removed in the latest version.

**Review by Proximyst — DISMISSED** [review]

**Review by aikar — DISMISSED** [review]

**Comment by Proximyst** [comment]
@limbo-app ready

**Review by limbo-app — APPROVED** [review]
Pull request automatically approved for merging - a core team member has previously approved this pull request.

**Pull request merged**
